New @DefineIn annotation for dynamic column list definition

mdhausman opened the pull request:
Added DefineIn annotation to the jdbi repository, created TestDefineInParameter.java for unit testing it.

This is for issue #512.

mdhausman commented:
@qualidafial Not sure why that test is failing with Travis. It works on my local dev environment. Might be worth noting that I'm using OpenJDK 1.8.0_91.

qualidafial commented on sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDefineInParameter.java:
```diff
@@ -0,0 +1,154 @@
+        @SqlUpdate("insert into <table> (<columns>) values (<values>)")
+        void unsupportedInsert(@Define("table") String table, @DefineIn List<String> columns, @BindIn("values") List<Object> values);
+
+        @SqlQuery("select <columns> from <table> where id = :id")
```
❤️

mdhausman replied:
Any idea why that one was failing in Travis?

qualidafial replied:
Your branch is a few commits behind `jdbi3`. I get the same error on my mac.

Try `git pull upstream jdbi3`

qualidafial commented on sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDefineInParameter.java:
```diff
@@ -0,0 +1,154 @@
+        values.add(1);
+        values.add("Some Pig");
+        testDao.badInsert("test", columnsArray, values);
+    }
```
Actually, I'd like array support. We just added it to `@BindIn` I want to say last week.

mdhausman replied:
Can-do

qualidafial commented:
@mdhausman Thanks so much for running the ball on this feature! I've been working on something today that could benefit from this feature. Very much looking forward to using this.

mdhausman commented on .gitignore:
```diff
@@ -20,3 +20,4 @@ test-output
 dependency-reduced-pom.xml
 
 .factorypath
+.reviewboardrc
```
😅 I totally forgot I could do that.

qualidafial commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/customizers/DefineIn.java:
```diff
@@ -0,0 +1,102 @@
+                } else {
+                    sb.append(',');
+                }
+                sb.append(o.toString());
```
Eh.. I think I'm coming round to throwing an exception on null. If somebody finds a use for it in the future, we can always add an annotation property to allow nulls and/or provide a substitute for nulls.

mdhausman replied:
Any strong preference as to whether it should be an IllegalArgument or an UnsupportedOperation?

qualidafial replied:
Agreed, IllegalArgumentException is probably better.

mdhausman commented:
Awesome, thanks for the feedback. This is all good stuff.

I'll get a new commit up as work permits within the week.

qualidafial commented on sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestDefineInParameter.java:
```diff
@@ -0,0 +1,154 @@
+        void badInsert(@Define("table") String table, @DefineIn("columns") Object columns, @BindIn("values") List<Object> values);
+
+        @SqlUpdate("insert into <table> (<columns>) values (<values>)")
+        void unsupportedInsert(@Define("table") String table, @DefineIn List<String> columns, @BindIn("values") List<Object> values);
```
Now that I look at it.. why _do_ we expect this to throw `UnsupportedOperationException`? I don't see anything obviously wrong with it.

mdhausman replied:
I guess I had thought that was the expected behavior of @Define, but I was misunderstanding the way Parameter works. Not used to working in Java 8! 

Now that I look at it again, it should only be happening when it's passed a raw value rather than a named variable. I have to wonder why this test was passing on my machine.

This will still be useful with some modification though, since it will be accepting arrays on the next iteration.

qualidafial replied:
The practical implication of this is that, if you compile your projects with `-parameters` set, then you can use bare `@Bind`, `@Define`, `@BindIn`, and now `@DefineIn` without declaring the bound name--and it will default to the parameter name.

mdhausman replied:
This is going to be very good to know when I start migrating my team's codebase to Java 8 in the not-so-distant future. Thanks for the breakdown!

qualidafial replied:
So probably the failing test just needs to be renamed to indicate that we're testing defaulting to the parameter name, and remove the expected exception.

qualidafial commented on core/src/main/java/org/jdbi/v3/core/Handle.java:
```diff
@@ -282,7 +282,7 @@ public int insert(String sql, Object... args) {
      */
     public int update(String sql, Object... args) {
-        Update stmt = createStatement(sql);
+        Update stmt = createUpdate(sql);
```
Something's screwy--this change came from upstream, it shouldn't be showed as a difference here. 😕

mdhausman replied:
Yeah, the missing upstream commits I pulled in appear to be rolled in here. Sorry! Still getting used to GitHub's PR system.

qualidafial commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/customizers/DefineIn.java:
```diff
@@ -0,0 +1,120 @@
+            if (arg instanceof List) {
+                argsList = (List<?>) arg;
+            } else if (arg instanceof Object[]) {
+                argsList = Arrays.asList((Object[]) arg);
```
Would be cool to support primitive arrays too, but we can do that in a follow-up ticket.

qualidafial commented:
This is looking really great.

If you could do me one last favor: try another `git pull upstream jdbi3`? It looks like your merge commit didn't mark the branch changes as reconciled--like your branch is duplicating the recent changes from upstream, instead of inheriting them.

qualidafial commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/customizers/DefineIn.java:
```diff
@@ -26,15 +26,29 @@
+ * {@code List} or {@code null} members in the {@code List} will result in an {@link IllegalArgumentException}.
+ * <p>
+ * Be aware of the list members you're binding with @DefineIn, as there is no input sanitization! <b>Blindly passing
+ * Strings through <code>@DefineIn</code> may make your application vulnerable to SQL Injection.</b>
```
👍

mdhausman commented:
I'm not sure that I can. At least, not without nuking the remote branch and creating a new PR.

```
♥ mhausman:~/opensource_code/jdbi (define_in_annotation) ‽ git push origin define_in_annotation
To https://github.com/wayfair/jdbi.git
 ! [rejected]        define_in_annotation -> define_in_annotation (non-fast-forward)
error: failed to push some refs to 'https://github.com/wayfair/jdbi.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
♥ mhausman:~/opensource_code/jdbi (define_in_annotation) ‽ git pull origin define_in_annotation 
From https://github.com/wayfair/jdbi
 * branch            define_in_annotation -> FETCH_HEAD
First, rewinding head to replay your work on top of it...
♥ mhausman:~/opensource_code/jdbi (define_in_annotation) ‽ git pull upstream jdbi3
From https://github.com/jdbi/jdbi
 * branch            jdbi3      -> FETCH_HEAD
First, rewinding head to replay your work on top of it...
Applying: Added DefineIn annotation to the jdbi repository, created TestDefineInParameter.java for unit testing it.
Applying: Fixed the gitignore, tightened up @DefineIn behavior and testing as requested in PR #528
♥ mhausman:~/opensource_code/jdbi (define_in_annotation) ‽ git push origin define_in_annotation 
Username for 'https://github.com': mdhausman
Password for 'https://mdhausman@github.com': 
To https://github.com/wayfair/jdbi.git
 ! [rejected]        define_in_annotation -> define_in_annotation (non-fast-forward)
error: failed to push some refs to 'https://github.com/wayfair/jdbi.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

qualidafial commented:
Huh. Maybe `git pull --rebase origin define_in_annotation`?

mdhausman commented:
So I actually already had `pull.rebase=true` configured globally. So I tried the opposite: `git pull --no-rebase`.

This has made things worse and I regret this decision. :sweat: 

I'm just nuking the branch and pushing it fresh.

mdhausman commented:
Not my finest hour, but the new one is up at #533.

stevenschlansker commented:
By the way, you can always fix the branch up locally and then force-push over this remote branch.  So it shouldn't be necessary to recreate the PR generally (it loses our discussion history)

mdhausman commented:
Oh. Well, now I know, and knowing is half the battle (GI Joooooeee). The other half is already lost because I already nuked it. Sorry! :sweat_smile: 

I've left a link forward from here and backward from the new one, which is the best I can do to keep the discussion history at this point.
